Add beanhub-cli and deps

fangpenlin opened the pull request:
## Description of changes

[beanhub-cli](https://github.com/LaunchPlatform/beanhub-cli) brings many useful beancount command tools powered by:
- [beanhub-forms](https://github.com/LaunchPlatform/beanhub-forms)
- [beanhub-import](https://github.com/LaunchPlatform/beanhub-import)
- [beanhub-extract](https://github.com/LaunchPlatform/beanhub-extract)

depends on: #310502

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

fangpenlin committed:
beanhub-extract: init at 0.0.7

fangpenlin committed:
beanhub-import: init at 0.1.5

fangpenlin committed:
beanhub-forms: init at 0.1.3

fangpenlin committed:
beanhub-cli: init at 1.1.3

kirillrdy commented:
if its still WIP, can you Convert to Draft so that people don't get notified

fangpenlin commented:
@kirillrdy hey, sorry. next time I will open it as DRAFT if it's still working in progress. It was waiting #310502 to merged as this PR depends on it. Now the PR was merged, I just rebased and pushed. This PR is ready for review.

ambroisie commented on pkgs/development/python-modules/beanhub-cli/default.nix:
```diff
@@ -0,0 +1,77 @@
+, fetchFromGitHub
+, buildPythonPackage
+, pythonOlder
+, pytz
+, pydantic
+, pydantic-core
+, pydantic-settings
+, pyyaml
+, jinja2
+, rich
+, fastapi
+, click
+, uvicorn
+, starlette-wtf
+, beancount-parser
+, beancount-black
+, beanhub-import
+, beanhub-forms
+, poetry-core
+, httpx
+, pytestCheckHook
```
I'd order the python modules ASCII-betically, and keep `pytestCheckHook` with the other packaging stuff.

ambroisie replied:
```suggestion
, pytestCheckHook
, beancount-black
, beancount-parser
, beanhub-forms
, beanhub-import
, click
, fastapi
, httpx
, jinja2
, poetry-core
, pydantic
, pydantic-core
, pydantic-settings
, pytz
, pyyaml
, rich
, starlette-wtf
, uvicorn
```

fangpenlin replied:
@ambroisie hey, thanks! sorry I am pretty new to contributing to nixpkgs. I usually run

```
nix fmt .
```

for my own nix projects. I just tried to run it but it says

```
error: flake 'git+file:///home/fangpen/workspace/playground/nixpkgs' does not provide attribute 'formatter.x86_64-linux'
```

ideally I think there should be a tool for this. but if not it's okay, thanks for calling it out anyway 🙇‍♀️

ambroisie commented on pkgs/development/python-modules/beanhub-cli/default.nix:
```diff
@@ -0,0 +1,77 @@
+    hash = "sha256-vYBbaUVJAs+aIp6aQpJb62DEDxe/sQTzgOkjPq6ADoc=";
+  };
+
+  buildInputs = [
+    poetry-core
+  ];
```
The input names seem to have changed, if I read [the docs](https://nixos.org/manual/nixpkgs/unstable/#buildpythonpackage-function) correctly this should be `build-system`.

ambroisie replied:
```suggestion
  build-system = [
    poetry-core
  ];
```

ambroisie commented on pkgs/development/python-modules/beanhub-cli/default.nix:
```diff
@@ -0,0 +1,77 @@
+    poetry-core
+  ];
+
+  propagatedBuildInputs = [
+    pytz
+    pydantic
+    pydantic-core
+    pydantic-settings
+    pyyaml
+    jinja2
+    rich
+    starlette-wtf
+    fastapi
+    click
+    uvicorn
+    beancount-parser
+    beancount-black
+    beanhub-import
+    beanhub-forms
+  ];
```
Similarly this should be `dependencies`.

Also please sort the list.

ambroisie replied:
```suggestion
  dependencies = [
    beancount-black
    beancount-parser
    beanhub-forms
    beanhub-import
    click
    fastapi
    jinja2
    pydantic
    pydantic-core
    pydantic-settings
    pytz
    pyyaml
    rich
    starlette-wtf
    uvicorn
  ];
```

ambroisie commented on pkgs/development/python-modules/beanhub-cli/default.nix:
```diff
@@ -0,0 +1,77 @@
+    homepage = "https://github.com/LaunchPlatform/beanhub-cli/";
+    changelog = "https://github.com/LaunchPlatform/beanhub-cli/releases/tag/${version}";
+    license = with licenses; [ mit ];
+    maintainers = with maintainers; [ ambroisie ];
```
I don't want to maintain this package, I don't make use of it.

Please add yourself to the maintainer list (in a commit named `maintains: add <handle>` beforein accordance with the [contributing docs](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) :-).

fangpenlin replied:
ah, sorry, didn't meant to throw these packages at your, forgot to add myself as an author 😅

ambroisie commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1461,6 +1461,14 @@ self: super: with self; {
 
   beancount-parser = callPackage ../development/python-modules/beancount-parser { };
 
+  beanhub-extract = callPackage ../development/python-modules/beanhub-extract { };
+
+  beanhub-import = callPackage ../development/python-modules/beanhub-import { };
+
+  beanhub-forms = callPackage ../development/python-modules/beanhub-forms { };
+
+  beanhub-cli = callPackage ../development/python-modules/beanhub-cli { };
+
```
Please sort these.

ambroisie commented on pkgs/development/python-modules/beanhub-extract/default.nix:
```diff
@@ -0,0 +1,47 @@
+, fetchFromGitHub
+, buildPythonPackage
+, pythonOlder
+, pytz
+, poetry-core
+, pytestCheckHook
```
Sorting

ambroisie commented on pkgs/development/python-modules/beanhub-extract/default.nix:
```diff
@@ -0,0 +1,47 @@
+    hash = "sha256-Wt8ZCyCaERNXEd0/Q89QWUW/wGFSHAP2RZLhnv5xkgY=";
+  };
+
+  buildInputs = [
+    poetry-core
+  ];
```
```suggestion
  build-system = [
    poetry-core
  ];
```

ambroisie commented on pkgs/development/python-modules/beanhub-forms/default.nix:
```diff
@@ -0,0 +1,55 @@
+    hash = "sha256-313c+ENmTe1LyfEiMXNB9AUoGx3Yv/1D0T3HnAbd+Zw=";
+  };
+
+  buildInputs = [
+    poetry-core
+  ];
```
```suggestion
  build-system = [
    poetry-core
  ];
```

ambroisie commented on pkgs/development/python-modules/beanhub-import/default.nix:
```diff
@@ -0,0 +1,59 @@
+    hash = "sha256-BH+d7cKC83b7T2Z1Xlwpsv3KACo1vDn6v65tElhir/c=";
+  };
+
+  buildInputs = [
+    poetry-core
+  ];
```
```suggestion
  build-system = [
    poetry-core
  ];
```

ambroisie commented on pkgs/development/python-modules/beanhub-extract/default.nix:
```diff
@@ -0,0 +1,47 @@
+    poetry-core
+  ];
+
+  propagatedBuildInputs = [
+    pytz
+  ];
```
```suggestion
  dependencies = [
    pytz
  ];
```

ambroisie commented on pkgs/development/python-modules/beanhub-forms/default.nix:
```diff
@@ -0,0 +1,55 @@
+    poetry-core
+  ];
+
+  propagatedBuildInputs = [
+    pydantic
+    pyyaml
+    wtforms
+    jinja2
+  ];
```
```suggestion
  dependencies = [
    jinja2
    pydantic
    pyyaml
    wtforms
  ];
```

ambroisie commented on pkgs/development/python-modules/beanhub-forms/default.nix:
```diff
@@ -0,0 +1,55 @@
+, fetchFromGitHub
+, buildPythonPackage
+, pythonOlder
+, pydantic
+, pyyaml
+, wtforms
+, jinja2
+, multidict
+, poetry-core
+, pytestCheckHook
```
```suggestion
, pytestCheckHook
, jinja2
, multidict
, poetry-core
, pydantic
, pyyaml
, wtforms
```

ambroisie commented on pkgs/development/python-modules/beanhub-import/default.nix:
```diff
@@ -0,0 +1,59 @@
+, fetchFromGitHub
+, buildPythonPackage
+, pythonOlder
+, pytz
+, pydantic
+, pyyaml
+, jinja2
+, beancount-parser
+, beancount-black
+, beanhub-extract
+, poetry-core
+, pytestCheckHook
```
```suggestion
, pytestCheckHook
, beancount-black
, beancount-parser
, beanhub-extract
, jinja2
, poetry-core
, pydantic
, pytz
, pyyaml
```

ambroisie commented on pkgs/development/python-modules/beanhub-import/default.nix:
```diff
@@ -0,0 +1,59 @@
+    poetry-core
+  ];
+
+  propagatedBuildInputs = [
+    pytz
+    pydantic
+    pyyaml
+    jinja2
+    beancount-parser
+    beancount-black
+    beanhub-extract
+  ];
```
```suggestion
  dependencies = [
    beancount-black
    beancount-parser
    beanhub-extract
    jinja2
    pydantic
    pytz
    pyyaml
  ];
```

ambroisie commented on pkgs/development/python-modules/beanhub-import/default.nix:
```diff
@@ -0,0 +1,59 @@
+  ];
+
+  meta = with lib; {
+    description = "Beanhub-import is a simple, declarative, smart, and easy-to-use library for importing extracted transactions from beanhub-extract. It generates Beancount transactions based on predefined rules.";
```
Nit: this description is a bit long IMO.

ambroisie reviewed (changes requested):
Thanks for adding the packages, a few nits mostly regarding evolutions `buildPythonPackage`'s API and general formatting.

fangpenlin committed:
maintainers: add fangpen

fangpenlin commented:
@ambroisie hey, sorry for pulling you into this unrelative PR, thanks again for the detailed review. I believe I have addressed all the feedbacks. Please let me know if there's anything else I can improve 🙏

kirillrdy commented on pkgs/development/python-modules/beanhub-cli/default.nix:
```diff
@@ -0,0 +1,77 @@
+buildPythonPackage rec {
+  pname = "beanhub-cli";
+  version = "1.1.3";
+  format = "pyproject";
```
```suggestion
  pyproject = true'
```

kirillrdy commented on pkgs/development/python-modules/beanhub-extract/default.nix:
```diff
@@ -0,0 +1,47 @@
+buildPythonPackage rec {
+  pname = "beanhub-extract";
+  version = "0.0.7";
+  format = "pyproject";
```
```suggestion
  pyproject = true;
```

kirillrdy commented on pkgs/development/python-modules/beanhub-forms/default.nix:
```diff
@@ -0,0 +1,55 @@
+buildPythonPackage rec {
+  pname = "beanhub-forms";
+  version = "0.1.3";
+  format = "pyproject";
```
```suggestion
  pyproject = true;
```

kirillrdy commented on pkgs/development/python-modules/beanhub-import/default.nix:
```diff
@@ -0,0 +1,59 @@
+buildPythonPackage rec {
+  pname = "beanhub-import";
+  version = "0.1.5";
+  format = "pyproject";
```
```suggestion
  pyproject = true;
```

kirillrdy commented on pkgs/development/python-modules/beanhub-cli/default.nix:
```diff
@@ -0,0 +1,77 @@
+    "beanhub_cli"
+  ];
+
+  meta = with lib; {
```
please consider adding https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

kirillrdy commented on pkgs/development/python-modules/beanhub-extract/default.nix:
```diff
@@ -0,0 +1,47 @@
+    "beanhub_extract"
+  ];
+
+  meta = with lib; {
```
https://github.com/NixOS/nixpkgs/issues/208242

ambroisie replied:
I agree with Sandro here, it's fine if it's limited to a small scope (here `meta`, but I also find it okay for `options`). 👎 on changing it IMO.

kirillrdy replied:
I don't have strong opinion here. I merely presenting people with information

kirillrdy reviewed (commented):
Result of `nixpkgs-review pr 310655` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python312Packages.beanhub-cli</li>
    <li>python312Packages.beanhub-cli.dist</li>
    <li>python312Packages.beanhub-import</li>
    <li>python312Packages.beanhub-import.dist</li>
  </ul>
</details>
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>beanhub-cli (python311Packages.beanhub-cli)</li>
    <li>beanhub-cli.dist (python311Packages.beanhub-cli.dist)</li>
    <li>python311Packages.beanhub-extract</li>
    <li>python311Packages.beanhub-extract.dist</li>
    <li>python311Packages.beanhub-forms</li>
    <li>python311Packages.beanhub-forms.dist</li>
    <li>python311Packages.beanhub-import</li>
    <li>python311Packages.beanhub-import.dist</li>
    <li>python312Packages.beanhub-extract</li>
    <li>python312Packages.beanhub-extract.dist</li>
    <li>python312Packages.beanhub-forms</li>
    <li>python312Packages.beanhub-forms.dist</li>
  </ul>
</details>

fangpenlin commented:
thanks for the review feedbacks! I will address them later today

fangpenlin commented:
@kirillrdy all the feedbacks addressed, please have a look again and let me know if there's anything else I can improve 🙏

kirillrdy reviewed (approved):
Result of `nixpkgs-review pr 310655` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>beanhub-cli (python311Packages.beanhub-cli)</li>
    <li>beanhub-cli.dist (python311Packages.beanhub-cli.dist)</li>
    <li>python311Packages.beanhub-extract</li>
    <li>python311Packages.beanhub-extract.dist</li>
    <li>python311Packages.beanhub-forms</li>
    <li>python311Packages.beanhub-forms.dist</li>
    <li>python311Packages.beanhub-import</li>
    <li>python311Packages.beanhub-import.dist</li>
    <li>python312Packages.beanhub-cli</li>
    <li>python312Packages.beanhub-cli.dist</li>
    <li>python312Packages.beanhub-extract</li>
    <li>python312Packages.beanhub-extract.dist</li>
    <li>python312Packages.beanhub-forms</li>
    <li>python312Packages.beanhub-forms.dist</li>
    <li>python312Packages.beanhub-import</li>
    <li>python312Packages.beanhub-import.dist</li>
  </ul>
</details>

fangpenlin commented:
@ambroisie can you please review again when you have time? 🙏
